### PR TITLE
[receiver/filelogreceiver] Add maxSurgeSize

### PR DESCRIPTION
**Pull request by Dainerx**
**Description:** <Describe what has changed.>

Adding a maximum surge size parameter to the stanza/fileconsumer reader. The primary goal is to prevent lagging behind a log file whose most recent entries have grown beyond the `maxSurgeSize`. For example, if 5 MB of data have been written to the log since the last read and the `maxSurgeSize` is set to 5 MB, we want to skip ahead and set the offset to the end of the file. See the image below for reference

offsetmaxsurgesize

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** 

2 Unit tests in stanza/fileconsumer reader

**Documentation:** <Describe the documentation added.>

Extended the readme of filelogreceiver

Questions: 
- Would this affect the reading mechanism when the file is compressed? 
- Should I add this behind feature gate?

### CODE REVIEW

File: receiver/filelogreceiver/README.md [hunk 1]
```diff
@@ -36,6 +36,7 @@ Tails and parses logs from files.
...
 | `poll_interval`                       | 200ms                                | The [duration](#time-parameters) between filesystem polls.                                                                                                                                                                                                      |
 | `fingerprint_size`                    | `1kb`                                | The number of bytes with which to identify a file. The first bytes in the file are used as the fingerprint. Decreasing this value at any point will cause existing fingerprints to forgotten, meaning that all files will be read from the beginning (one time) |
+| `max_surge_size`                      |                                      | The maximum surge size of a log file to read. log entries will be skipped if it is larger than `max_surge_size`. Protects against lagging behind when reading large log files.                                                                                  |
```
Comment (rafaaseddik) [inline receiver/filelogreceiver/README.md:39]: It doesn't have a default value, or is it missing documentation ?

Comment (Dainerx) [reply receiver/filelogreceiver/README.md:39]: default value is zero value, will update

**Comment by djaglowski** [comment]
If I'm understanding correctly, the idea is to provide a mechanism which would trigger skipping over logs to reach the end of the file immediately. Can you give me more details about this use case where this is needed?

**Comment by Dainerx** [comment]
@djaglowski yes correct. 

A use case is as follow: 

Given a filelogreceiver with poll_interval of **10s** and max_surge_size of **10Mb,** watching `myservice.log`. 

t1=10s: 2Kb were written // ok
t2=10s: 5Kb were written // ok
t3=10s: 20Mb were written // This is too much log data and we want to skip it
t4 = 10s: 1Mb were written // ok

**Comment by djaglowski** [comment]
> @djaglowski yes correct.
> 
> A use case is as follow:
> 
> Given a filelogreceiver with poll_interval of **10s** and max_surge_size of **10Mb,** watching `myservice.log`.
> 
> t1=10s: 2Kb were written // ok t2=10s: 5Kb were written // ok t3=10s: 20Mb were written // This is too much log data and we want to skip it t4 = 10s: 1Mb were written // ok

I would like to understand the _reason_ why there is too much log data. Typically a burst of data could indicate a problem, which I would think users want to understand fully. Alternately they can tune down the amount of logs generated or right size the collector to consume more data and then filter out that which isn't interesting.

**Comment by overmeulen** [comment]
Hello @djaglowski 
This is a protection mechanism.
We had the use case in the past of specific log files increasing by hundreds of MB in a few seconds, because of wrongly configured loggers or badly implemented retry mechanisms...
We know this will happen again.
This optional feature aims at protecting the collector from this use case, in terms of CPU, memory and latencies as we target a near real time monitoring

File: pkg/stanza/fileconsumer/internal/reader/reader.go [hunk 2]
```diff
@@ -98,6 +99,21 @@ func (r *Reader) ReadToEnd(ctx context.Context) {
 		}
 	}()
 
+	if r.maxSurgeSize > 0 {
+		info, _ := r.file.Stat()
+		currentEOF := info.Size()
+		toEOF := currentEOF - r.Offset
+		if toEOF >= r.maxSurgeSize {
+			r.set.Logger.Debug("file too large, moving offset to the end of the file", zap.Int64("size", toEOF), zap.Int64("maxSurgeSize", r.maxSurgeSize))
```
Comment (overmeulen) [inline pkg/stanza/fileconsumer/internal/reader/reader.go:107]: It's not the file that is too large, it's the amount of new logs

File: receiver/filelogreceiver/README.md [hunk 3]
```diff
@@ -36,6 +36,7 @@ Tails and parses logs from files.
...
 | `poll_interval`                       | 200ms                                | The [duration](#time-parameters) between filesystem polls.                                                                                                                                                                                                      |
 | `fingerprint_size`                    | `1kb`                                | The number of bytes with which to identify a file. The first bytes in the file are used as the fingerprint. Decreasing this value at any point will cause existing fingerprints to forgotten, meaning that all files will be read from the beginning (one time) |
+| `max_surge_size`                      | 0                                    | The maximum surge size of a log file to read. log entries will be skipped if it is larger than `max_surge_size`. Protects against lagging behind when reading large log files.                                                                                  |
```
Comment (overmeulen) [inline receiver/filelogreceiver/README.md:39]: This description is not very clear for me

**Comment by djaglowski** [comment]
@overmeulen, thanks for explaining further. I can appreciate the motivation but I'm not convinced this is the approach we should provide. 

Here are a few reasons I don't like the idea of skipping directly to the end of a file:
1. Skipping directly to the end of file provides no guarantee that the new starting point aligns with the start of a log. This feels quite crude and may have unintended impacts downstream.
2. This approach also feels crude in the sense that even if we were to skip ahead, wouldn't the user still want to read some of the most recent logs?
3. File rotation settings can be used to constrain the amount of data preserved. Even if they are not used specifically for this purpose, skipping to the end of a file can easily be ineffective because of file rotation settings. For example, if your skip ahead limit is 5 MB, then any file rotation config which rotates at less than 5 MB can still easily bypass the protective benefit of this design. e.g. 100 x 1 MB files will not be limited by this mechanism.
4. Often when there is a surge of logs it is the result of a problem occurring, in which case this is exactly the scenario where I would most want to ensure all logs are delivered. Constraining the load this places on a system is an orthogonal concern. 
5. I can't find any other examples of logging agents which rely on this mechanism. I've checked FluentBit, vector, elastic beats, but please let me know if I missed an example.

Ultimately, I would be much more receptive to placing limits on the number or total size of logs read during a poll interval. This should effectively limit the load placed on the system when processing bursts while also avoiding loss of data. If there is a situation where this collector _generally_ cannot keep up with the amount of logs written, that is a sizing issue and I don't believe we should try to solve it here.

**Comment by Dainerx** [comment]
__
 
> Ultimately, I would be much more receptive to placing limits on the number or total size of logs read during a poll interval. 
Setting a limit for the total entries to be read in one poll, will just leave the entries for the next poll,  . Would you accept a warning when we the limit is reached?

**Comment by djaglowski** [comment]
> Hence, you'll lag behind ever further.

Right, but this is an indication of an ongoing sizing issue, not a one-off exception that we should handle. 

> Would you accept just a warning when we the limit is reached?

How about a warning that prints when a "max logs per poll" limit is reached?

**Pull request closed without merging**